### PR TITLE
add pprof endpoint

### PR DESCRIPTION
**Pull request by zhangzhenhao**
this pr is try to implement #369 

@Random-Liu  I have try to implement it, Please have a look.

Signed-off-by: zhangzhenhao <zhangzhenhao@outlook.com>

### CODE REVIEW

File: cmd/cri-containerd/options/options.go [hunk 1]
```diff
@@ -90,6 +90,12 @@ type Config struct {
 	SystemdCgroup bool `toml:"systemd_cgroup"`
 	// OOMScore adjust the cri-containerd's oom score
 	OOMScore int `toml:"oom_score"`
+	// EnableProfiling is used for Enable profiling via host:port/debug/pprof/
```
[the remark this answers is not in the document]
Comment (zhangzhenhao) [reply cmd/cri-containerd/options/options.go]: @yanxuean  thanks a lot for your review. I will fix this issue.

File: cmd/cri-containerd/options/options.go [hunk 2]
```diff
@@ -90,6 +90,12 @@ type Config struct {
...
+	EnableProfiling bool `toml:"profiling"`
+	// ProfilingPort is the port for profiling via host:port/debug/pprof/
+	ProfilingPort string `toml:"profiling-port"`
```
[the remark this answers is not in the document]
Comment (zhangzhenhao) [reply cmd/cri-containerd/options/options.go]: @yanxuean  thanks a lot for your review. I will fix this issue.

File: pkg/server/service.go [hunk 3]
```diff
@@ -268,3 +276,13 @@ func (c *criContainerdService) getDeviceUUID(path string) (string, error) {
 func imageFSPath(rootDir, snapshotter string) string {
 	return filepath.Join(rootDir, fmt.Sprintf("%s.%s", plugin.SnapshotPlugin, snapshotter))
 }
+
+// startProfilingServer start http server to profiling via web interface
+func startProfilingServer(addr string, port string) {
+	if isIP := (net.ParseIP(addr) != nil); isIP {
```
[the remark this answers is not in the document]
Comment (zhangzhenhao) [reply pkg/server/service.go]: @yanxuean  thanks a lot for your review. I will do as your suggestion.

File: pkg/server/service.go [hunk 4]
```diff
@@ -268,3 +276,13 @@ func (c *criContainerdService) getDeviceUUID(path string) (string, error) {
...
+		endpoint := net.JoinHostPort(addr, port)
+		http.ListenAndServe(endpoint, nil)
+	} else {
+		glog.Warningf("Ignore Start Profiling Server for invalid IP Address:%q", addr)
```
[the remark this answers is not in the document]
Comment (zhangzhenhao) [reply pkg/server/service.go]: Warning is for this server is not necessary for cri-containerd can work. I am not sure about that.

File: pkg/server/service.go [hunk 5]
```diff
@@ -268,3 +276,13 @@ func (c *criContainerdService) getDeviceUUID(path string) (string, error) {
 func imageFSPath(rootDir, snapshotter string) string {
 	return filepath.Join(rootDir, fmt.Sprintf("%s.%s", plugin.SnapshotPlugin, snapshotter))
 }
+
+// startProfilingServer start http server to profiling via web interface
+func startProfilingServer(addr string, port string) {
+	if isIP := (net.ParseIP(addr) != nil); isIP {
+		endpoint := net.JoinHostPort(addr, port)
+		http.ListenAndServe(endpoint, nil)
```
[the remark this answers is not in the document]
Comment (zhangzhenhao) [reply pkg/server/service.go]: thanks for your review, I will use the same log level as check of IP.

File: cmd/cri-containerd/options/options.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (zhangzhenhao) [reply cmd/cri-containerd/options/options.go]: Done

File: cmd/cri-containerd/options/options.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (zhangzhenhao) [reply cmd/cri-containerd/options/options.go]: Done

File: pkg/server/service.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (zhangzhenhao) [reply pkg/server/service.go]: Done

File: pkg/server/service.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (zhangzhenhao) [reply pkg/server/service.go]: Thanks a lot for your review. 
Done

File: pkg/server/service.go [hunk 6]
```diff
@@ -268,3 +276,21 @@ func (c *criContainerdService) getDeviceUUID(path string) (string, error) {
...
+	if ip := net.ParseIP(addr); ip != nil {
+		endpoint := net.JoinHostPort(addr, port)
+		mux := http.NewServeMux()
+		mux.HandleFunc("/debug/pprof/", pprof.Index)
+		mux.HandleFunc("/debug/pprof/profile", pprof.Profile)
+		mux.HandleFunc("/debug/pprof/symbol", pprof.Symbol)
+		mux.HandleFunc("/debug/pprof/trace", pprof.Trace)
+		err := http.ListenAndServe(endpoint, mux)
+		if err != nil {
+			glog.Errorf("Failed to start server: %v", err)
```
Comment (mikebrow) [inline pkg/server/service.go]: /s/server/profiling server/

File: pkg/server/service.go [hunk 7]
```diff
@@ -228,6 +230,12 @@ func (c *criContainerdService) Run() error {
 		close(grpcServerCloseCh)
 	}()
 
+	// Start profiling server if enable.
+	if c.config.EnableProfiling {
+		glog.V(2).Info("Start profiling server")
+		go startProfilingServer(c.config.ProfilingAddress, c.config.ProfilingPort)
+	}
```
Comment (mikebrow) [inline pkg/server/service.go]: Any ideas for testing if it's working?

File: cmd/cri-containerd/options/options.go [hunk 8]
```diff
@@ -213,5 +225,8 @@ func defaultConfig() Config {
 		StatsCollectPeriod:  10,
 		SystemdCgroup:       false,
 		OOMScore:            -999,
+		EnableProfiling:     true,
+		ProfilingPort:       "6060",
```
Comment (Random-Liu) [inline cmd/cri-containerd/options/options.go]: Where did you get this number? We should avoid port conflict with other system daemons.

We once have a production issue that kube-proxy has conflict port with node-problem-detector.

File: pkg/server/service.go [hunk 9]
```diff
@@ -268,3 +276,21 @@ func (c *criContainerdService) getDeviceUUID(path string) (string, error) {
...
+		mux.HandleFunc("/debug/pprof/", pprof.Index)
+		mux.HandleFunc("/debug/pprof/profile", pprof.Profile)
+		mux.HandleFunc("/debug/pprof/symbol", pprof.Symbol)
+		mux.HandleFunc("/debug/pprof/trace", pprof.Trace)
+		err := http.ListenAndServe(endpoint, mux)
+		if err != nil {
+			glog.Errorf("Failed to start server: %v", err)
+		}
+	} else {
+		glog.Warningf("Ignore Start Profiling Server for invalid IP Address:%q", addr)
```
Comment (Random-Liu) [inline pkg/server/service.go]: Why capitalize?

File: pkg/server/service.go [hunk 10]
```diff
@@ -268,3 +276,21 @@ func (c *criContainerdService) getDeviceUUID(path string) (string, error) {
 func imageFSPath(rootDir, snapshotter string) string {
 	return filepath.Join(rootDir, fmt.Sprintf("%s.%s", plugin.SnapshotPlugin, snapshotter))
 }
+
+// startProfilingServer start http server to profiling via web interface
+func startProfilingServer(addr string, port string) {
```
Comment (Random-Liu) [inline pkg/server/service.go]:
``` golang
ip := net.ParseIP(addr)
if ip == nil {
  glog.Warningf("Ignore Start Profiling Server for invalid IP Address:%q", addr)
  return
}
// Do other thing
```
Less indent is always preferred.

Comment (Random-Liu) [reply pkg/server/service.go]: And why can't I use host name here?

File: pkg/server/service.go [hunk 11]
```diff
@@ -228,6 +230,12 @@ func (c *criContainerdService) Run() error {
 		close(grpcServerCloseCh)
 	}()
 
+	// Start profiling server if enable.
+	if c.config.EnableProfiling {
+		glog.V(2).Info("Start profiling server")
+		go startProfilingServer(c.config.ProfilingAddress, c.config.ProfilingPort)
```
Comment (Random-Liu) [inline pkg/server/service.go]: I feel like that we should move this to `main`. This is not business logic related, let's move it out of `pkg/server`. :)

File: pkg/server/service.go (hunk 6 above)
Comment (zhangzhenhao) [reply pkg/server/service.go]: thanks a lot for your review. I will do as your suggestion.

File: pkg/server/service.go (hunk 9 above)
Comment (zhangzhenhao) [reply pkg/server/service.go]: I am sorry for that. Will pay attention to this issue.

Comment (zhangzhenhao) [reply pkg/server/service.go]: BTW, do you think use "Error" is better than "Warning"?

File: pkg/server/service.go (hunk 11 above)
Comment (zhangzhenhao) [reply pkg/server/service.go]: Thanks for your review. will do as your suggestion..

File: pkg/server/service.go (hunk 9 above)
Comment (Random-Liu) [reply pkg/server/service.go]: Should be an error. Actually I don't even think we need to enforce this to be an ip. Hostname should also be ok.

File: cmd/cri-containerd/options/options.go (hunk 8 above)
Comment (zhangzhenhao) [reply cmd/cri-containerd/options/options.go]: I am sorry to say, I am not sure about  what  is better,  I use it for I have found cri-o use 6060 as port of pprof endpoint.  Maybe do not give default is better.  This problem is bothering me. Do you have any suggestions?

File: pkg/server/service.go (hunk 10 above)
Comment (zhangzhenhao) [reply pkg/server/service.go]: Only for check this value is vaiid . I will remove this limit.

Comment (zhangzhenhao) [reply pkg/server/service.go]: Done

File: pkg/server/service.go (hunk 9 above)
Comment (zhangzhenhao) [reply pkg/server/service.go]: Done

File: pkg/server/service.go (hunk 6 above)
Comment (zhangzhenhao) [reply pkg/server/service.go]: Done

File: pkg/server/service.go (hunk 11 above)
Comment (zhangzhenhao) [reply pkg/server/service.go]: Done

File: cmd/cri-containerd/options/options.go (hunk 8 above)
Comment (Random-Liu) [reply cmd/cri-containerd/options/options.go]: Let's use 10009 or 10011, which is around our stream server port.

**Comment by Random-Liu** [comment]
LGTM other than the port number.

File: cmd/cri-containerd/options/options.go (hunk 8 above)
Comment (zhangzhenhao) [reply cmd/cri-containerd/options/options.go]: @Random-Liu  Thanks a lot for your suggestion.  use 10011 as profiling port.

Comment (yanxuean) [reply cmd/cri-containerd/options/options.go]: The double 11 is coming.

**Comment by Random-Liu** [comment]
Hm. I tried squash and merge, and it seems that the merge commit is not genernated. Hope that doesn't hurt.

**Pull request merged**